lib/AST/Decl.cpp

```diff
@@ -445,6 +445,26 @@ case DeclKind::ID: return cast<ID##Decl>(this)->getLoc();
   llvm_unreachable("Unknown decl kind");
 }
 
+Expr *AbstractFunctionDecl::getSingleExpressionBody() const {
+  assert(hasSingleExpressionBody() && "Not a single-expression body");
+  auto braceStmt = getBody();
+  assert(braceStmt != nullptr && "No body currently available.");
+  auto body = getBody()->getElement(0);
+  if (body.is<Stmt *>())
+    return cast<ReturnStmt>(body.get<Stmt *>())->getResult();
```

Comment:
Nitpick (here and elsewhere): We'd usually write this as

```
if (auto *stmt = body.dyn_cast<Stmt *>())
  return cast<ReturnStmt>(stmt)->getResult();
```

to avoid the `is`/`get` repetition.

Reply:
Can we get into the state where it's an expression, though? It seems like the way you have your change set up, the parser will always produce a ReturnStmt.

Reply:
Fixed up all the casting to use `dyn_cast` instead of the repetitive use of `is` and `get` in 
https://github.com/apple/swift/pull/23251/commits/a310279dabda30f1de50b2a6bb2c11c0ef20868f .

Reply:
I think we can get into that state after the second modification of the body happens, because the return that got added gets removed again, right?

Reply:
At that point it's not a single-expression closure, right?

Reply:
I'm not sure.  Looking at the way that this is done for closures (`ExprWalker::walkToExprPre`, `ExprRewriter::coerceClosureExprFromNever` and `ExprRewriter::coerceClosureExprToVoid`), it looks like the closures remain single expression even though their bodies are rewritten not to have the return (in the case of `coerceClosureExprFromNever`) or two have the original expression followed by a `return ()` (in the case of `coerceClosureExprToVoid`).

Reply:
Ah, I see. The flag is more about how it was written in source, then. Okay, that makes sense.